File: azul_rl/game/factory.py
import random
from typing import Dict, List, Optional

from .tile import Tile, TileColor


class Factory:
    """Represents a factory display that holds 4 tiles."""

    def __init__(self):
        self.tiles: List[Tile] = []

    def fill_from_bag(self, bag: List[Tile]) -> None:
        """Fill factory with 4 tiles from the bag."""
        self.tiles = []
        for _ in range(4):
            if bag:
                tile = bag.pop()
                self.tiles.append(tile)

    def take_tiles(self, color: TileColor) -> tuple[List[Tile], List[Tile]]:
        """Take all tiles of specified color. Returns (taken_tiles, remaining_tiles)."""
        taken = []
        remaining = []

        for tile in self.tiles:
            if tile.color == color:
                taken.append(tile)
            else:
                remaining.append(tile)

        self.tiles = []
        return taken, remaining

    def is_empty(self) -> bool:
        """Check if factory is empty."""
        return len(self.tiles) == 0

    def has_color(self, color: TileColor) -> bool:
        """Check if factory has tiles of specified color."""
        return any(tile.color == color for tile in self.tiles)

    def get_available_colors(self) -> List[TileColor]:
        """Get list of colors available in this factory."""
        colors = set()
        for tile in self.tiles:
            if tile.color != TileColor.FIRST_PLAYER:
                colors.add(tile.color)
        return list(colors)

    def __repr__(self) -> str:
        return f"Factory({[str(tile) for tile in self.tiles]})"


class CenterArea:
    """Represents the center area where leftover tiles accumulate."""

    def __init__(self):
        self.tiles: List[Tile] = []
        self.has_first_player_marker = False

    def add_tiles(self, tiles: List[Tile]) -> None:
        """Add tiles to the center area."""
        self.tiles.extend(tiles)

    def add_first_player_marker(self) -> None:
        """Add the first player marker to the center."""
        self.has_first_player_marker = True

    def take_tiles(self, color: TileColor) -> List[Tile]:
        """Take all tiles of specified color from center."""
        taken = []
        remaining = []

        for tile in self.tiles:
            if tile.color == color:
                taken.append(tile)
            else:
                remaining.append(tile)

        self.tiles = remaining

        # If taking tiles from center for first time this round, also take first player marker
        if taken and self.has_first_player_marker:
            taken.append(Tile.create_first_player_marker())
            self.has_first_player_marker = False

        return taken

    def is_empty(self) -> bool:
        """Check if center area is empty (ignoring first player marker)."""
        return len(self.tiles) == 0

    def has_color(self, color: TileColor) -> bool:
        """Check if center has tiles of specified color."""
        return any(tile.color == color for tile in self.tiles)

    def get_available_colors(self) -> List[TileColor]:
        """Get list of colors available in center."""
        colors = set()
        for tile in self.tiles:
            if tile.color != TileColor.FIRST_PLAYER:
                colors.add(tile.color)
        return list(colors)

    def clear(self) -> None:
        """Clear the center area."""
        self.tiles = []
        self.has_first_player_marker = False

    def __repr__(self) -> str:
        tiles_str = [str(tile) for tile in self.tiles]
        if self.has_first_player_marker:
            tiles_str.append("first_player")
        return f"Center({tiles_str})"


class FactoryArea:
    """Manages all factories and the center area."""

    def __init__(self, num_players: int):
        # Number of factories = 2 * num_players + 1
        self.num_factories = 2 * num_players + 1
        self.factories: List[Factory] = [Factory() for _ in range(self.num_factories)]
        self.center = CenterArea()

    def setup_round(self, bag: List[Tile]) -> None:
        """Setup factories for a new round."""
        # Clear center area
        self.center.clear()

        # Add first player marker to center
        self.center.add_first_player_marker()

        # Fill each factory with 4 tiles
        for factory in self.factories:
            factory.fill_from_bag(bag)

    def take_from_factory(self, factory_index: int, color: TileColor) -> List[Tile]:
        """Take tiles from a specific factory."""
        if factory_index < 0 or factory_index >= len(self.factories):
            return []

        taken, remaining = self.factories[factory_index].take_tiles(color)

        # Add remaining tiles to center
        if remaining:
            self.center.add_tiles(remaining)

        return taken

    def take_from_center(self, color: TileColor) -> List[Tile]:
        """Take tiles from center area."""
        return self.center.take_tiles(color)

    def is_round_over(self) -> bool:
        """Check if all factories and center are empty."""
        all_factories_empty = all(factory.is_empty() for factory in self.factories)
        center_empty = self.center.is_empty()
        return all_factories_empty and center_empty

    def get_available_moves(self) -> List[tuple[int, TileColor]]:
        """Get all available moves. Returns list of (source, color) where source -1 = center."""
        moves = []

        # Check center
        for color in self.center.get_available_colors():
            moves.append((-1, color))

        # Check factories
        for i, factory in enumerate(self.factories):
            for color in factory.get_available_colors():
                moves.append((i, color))

        return moves

    def __repr__(self) -> str:
        return f"FactoryArea(factories={self.factories}, center={self.center})"

File: game_cpp/factory.cpp
#include "factory.h"
#include <algorithm>
#include <unordered_set>

namespace azul {

// Factory implementation
Factory::Factory() = default;

void Factory::fill_from_bag(std::vector<Tile>& bag) {
    tiles_.clear();
    for (int i = 0; i < 4 && !bag.empty(); ++i) {
        tiles_.push_back(bag.back());
        bag.pop_back();
    }
}

std::pair<std::vector<Tile>, std::vector<Tile>> Factory::take_tiles(TileColor color) {
    std::vector<Tile> taken;
    std::vector<Tile> remaining;
    
    for (const auto& tile : tiles_) {
        if (tile.color() == color) {
            taken.push_back(tile);
        } else {
            remaining.push_back(tile);
        }
    }
    
    tiles_.clear();
    return {taken, remaining};
}

auto Factory::is_empty() const -> bool {
    return tiles_.empty();
}

auto Factory::has_color(TileColor color) const -> bool {
    return std::any_of(tiles_.begin(), tiles_.end(),
                      [color](const Tile& tile) { return tile.color() == color; });
}

auto Factory::get_available_colors() const -> std::vector<TileColor> {
    std::vector<TileColor> colors;
    colors.reserve(5); // At most 5 different colors
    
    // Use simple vector with duplicate check - faster than unordered_set for small collections
    for (const auto& tile : tiles_) {
        if (tile.color() != TileColor::FIRST_PLAYER) {
            // Check if color already added (linear search is faster than hash for small vectors)
            bool found = false;
            for (const auto& existing_color : colors) {
                if (existing_color == tile.color()) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                colors.push_back(tile.color());
            }
        }
    }
    return colors;
}

auto Factory::copy() const -> Factory {
    Factory new_factory;
    new_factory.tiles_ = tiles_;
    return new_factory;
}

// CenterArea implementation
CenterArea::CenterArea() 
    : has_first_player_marker_(false), first_player_marker_(Tile::create_first_player_marker()) {}

void CenterArea::add_tiles(const std::vector<Tile>& tiles) {
    tiles_.insert(tiles_.end(), tiles.begin(), tiles.end());
}

void CenterArea::add_first_player_marker() {
    has_first_player_marker_ = true;
}

auto CenterArea::take_tiles(TileColor color) -> std::vector<Tile> {
    std::vector<Tile> taken;
    std::vector<Tile> remaining;
    
    for (const auto& tile : tiles_) {
        if (tile.color() == color) {
            taken.push_back(tile);
        } else {
            remaining.push_back(tile);
        }
    }
    
    tiles_ = remaining;
    
    // If taking tiles from center for first time this round,
    // also take first player marker
    if (!taken.empty() && has_first_player_marker_) {
        taken.push_back(first_player_marker_);
        has_first_player_marker_ = false;
    }
    
    return taken;
}

auto CenterArea::is_empty() const -> bool {
    return tiles_.empty();
}

auto CenterArea::has_color(TileColor color) const -> bool {
    return std::any_of(tiles_.begin(), tiles_.end(),
                      [color](const Tile& tile) { return tile.color() == color; });
}

auto CenterArea::get_available_colors() const -> std::vector<TileColor> {
    std::vector<TileColor> colors;
    colors.reserve(5); // At most 5 different colors
    
    // Use simple vector with duplicate check - faster than unordered_set for small collections
    for (const auto& tile : tiles_) {
        if (tile.color() != TileColor::FIRST_PLAYER) {
            // Check if color already added (linear search is faster than hash for small vectors)
            bool found = false;
            for (const auto& existing_color : colors) {
                if (existing_color == tile.color()) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                colors.push_back(tile.color());
            }
        }
    }
    return colors;
}

void CenterArea::clear() {
    tiles_.clear();
    has_first_player_marker_ = false;
}

auto CenterArea::copy() const -> CenterArea {
    CenterArea new_center;
    new_center.tiles_ = tiles_;
    new_center.has_first_player_marker_ = has_first_player_marker_;
    new_center.first_player_marker_ = first_player_marker_;
    return new_center;
}

// FactoryArea implementation
FactoryArea::FactoryArea(int num_players) 
    : num_factories_((2 * num_players) + 1), factories_(num_factories_) {}

void FactoryArea::setup_round(std::vector<Tile>& bag) {
    // Clear center area
    center_.clear();
    
    // Clear all factories first
    for (auto& factory : factories_) {
        factory = Factory(); // Reset to empty state
    }
    
    // Add first player marker to center
    center_.add_first_player_marker();
    
    // Fill each factory with 4 tiles
    for (auto& factory : factories_) {
        factory.fill_from_bag(bag);
    }
}

auto FactoryArea::take_from_factory(int factory_index, TileColor color) -> std::vector<Tile> {
    if (factory_index < 0 || factory_index >= static_cast<int>(factories_.size())) {
        return {};
    }
    
    auto [taken, remaining] = factories_[factory_index].take_tiles(color);
    
    // Add remaining tiles to center
    if (!remaining.empty()) {
        center_.add_tiles(remaining);
    }
    
    return taken;
}

auto FactoryArea::take_from_center(TileColor color) -> std::vector<Tile> {
    return center_.take_tiles(color);
}

auto FactoryArea::is_round_over() const -> bool {
    bool all_factories_empty = std::all_of(factories_.begin(), factories_.end(),
                                          [](const Factory& factory) { return factory.is_empty(); });
    
    // Center is considered empty if it has no regular tiles
    bool center_has_regular_tiles = std::any_of(center_.tiles().begin(), center_.tiles().end(),
                                               [](const Tile& tile) { return tile.color() != TileColor::FIRST_PLAYER; });
    
    return all_factories_empty && !center_has_regular_tiles;
}

auto FactoryArea::get_available_moves() const -> std::vector<std::pair<int, TileColor>> {
    std::vector<std::pair<int, TileColor>> moves;
    moves.reserve(50); // Reasonable estimate to avoid reallocations
    
    // Check center first (like Python)
    auto center_colors = center_.get_available_colors();
    for (const auto color : center_colors) {
        moves.emplace_back(-1, color);
    }
    
    // Check factories (like Python) - no need to check is_empty()
    for (int i = 0; i < static_cast<int>(factories_.size()); ++i) {
        auto factory_colors = factories_[i].get_available_colors();
        for (const auto color : factory_colors) {
            moves.emplace_back(i, color);
        }
    }
    
    return moves;
}

auto FactoryArea::copy() const -> FactoryArea {
    FactoryArea new_area(0); // Dummy num_players, will be overwritten
    new_area.num_factories_ = num_factories_;
    new_area.factories_.clear();
    new_area.factories_.reserve(factories_.size());
    for (const auto& factory : factories_) {
        new_area.factories_.push_back(factory.copy());
    }
    new_area.center_ = center_.copy();
    return new_area;
}

} // namespace azul 